### PR TITLE
Feature/update rise set

### PR DESCRIPTION
**Pull request by dmcollom**
Refactored rise/set and sunup calculations
Refactored updatestatus management command
Added tests

### CODE REVIEW

File: tom_observations/utils.py [hunk 1]
```diff
@@ -1,23 +1,117 @@
...
-        next_rising = datetime.strptime(str(ephem.date(observer.next_rising(target))), ephem_format)
-        next_setting = datetime.strptime(str(ephem.date(observer.next_setting(target))), ephem_format)
-        rise_set.add_rise_set_pair(RiseSetPair(next_rising, next_setting))
-        start_time = datetime.strptime(str(ephem.date(next_setting)), ephem_format) + timedelta(seconds=1)
-    return rise_set
\ No newline at end of file
+        next_setting = ephem_to_timestamp(observer.next_setting(target))
+        rise_set.append((previous_rising, next_setting))
+        start_time = next_setting + 1
+    while start_time < end_time.timestamp():
```
Comment (cmccully) [inline tom_observations/utils.py]: There are tests in the LCO rise_set library that might be worth checking against this.

File: tom_observations/utils.py [hunk 2]
```diff
@@ -1,23 +1,117 @@
...
+        time value used to find the most recent rise, in UNIX time
+
+    Returns
+    -------
+    tuple
+        Most recent rise/set pair with respect to the given time
+
+    """
+    last_rise_pos = bisect_left(rise_sets, (time,))
+    if last_rise_pos-1 < 0:
```
Comment (cmccully) [inline tom_observations/utils.py]: PEP8.

Comment (cmccully) [reply tom_observations/utils.py]: Also, why not if late_rise_pos <=0:

File: tom_targets/models.py [hunk 3]
```diff
@@ -132,20 +175,18 @@ def get_visibility(self, start_time, end_time, interval, airmass_limit=10):
...
-                    rise_set_for_sun = rise_sets.get_last_rise(datetime.fromtimestamp(time))
-                    sunup = datetime.fromtimestamp(time) < rise_set_for_sun.set if rise_set_for_sun else False
+                    last_rise_set = get_last_rise_set_pair(rise_sets, time)
+                    sunup = time > last_rise_set[0] and time < last_rise_set[1] if last_rise_set else False
                     observer.date = datetime.fromtimestamp(time)
                     body.compute(observer)
-                    alt = Angle(str(body.alt) + ' degrees')
-                    az = Angle(str(body.az) + ' degrees')
+                    alt = Angle(str(body.alt) + 'd')
+                    az = Angle(str(body.az) + 'd')
```
Comment (cmccully) [inline tom_targets/models.py]: can you do Angle with a units argument? It might be clearer.

File: tom_targets/tests/factories.py [hunk 4]
```diff
@@ -9,6 +9,7 @@ class Meta:
 
     identifier = factory.Faker('pystr')
     name = factory.Faker('pystr')
+    type = Target.SIDEREAL
```
Comment (cmccully) [inline tom_targets/tests/factories.py]: type is a reserved word in python.

File: tom_observations/utils.py [hunk 5]
```diff
@@ -1,23 +1,117 @@
...
+
+    Returns
+    -------
+    array
+        An array of tuples, each a pair of values representing a rise and a set,
+        both in UNIX time
+
+    """
+    if end_time < start_time:
+        raise Exception('Start must be before end')
```
[the remark this answers is not in the document]
Comment (dmcollom) [reply tom_observations/utils.py]: Fixed, everything is now using datetime objects.

File: tom_targets/models.py [hunk 6]
```diff
@@ -131,21 +174,22 @@ def get_visibility(self, start_time, end_time, interval, airmass_limit=10):
...
+                curr_interval = start_time
+                while curr_interval <= end_time:
+                    time = curr_interval
+                    last_rise_set = get_last_rise_set_pair(rise_sets, time)
+                    sunup = time > last_rise_set[0] and time < last_rise_set[1] if last_rise_set else False
+                    observer.date = curr_interval
                     body.compute(observer)
-                    alt = Angle(str(body.alt) + ' degrees')
-                    az = Angle(str(body.az) + ' degrees')
+                    alt = Angle(str(body.alt), unit=units.degree)
```
Comment (cmccully) [inline tom_targets/models.py]: I don't think you need to cast to strings here now do you?

Comment (dmcollom) [reply tom_targets/models.py]: body.alt returns a pyephem degrees object, so unfortunately, I do.

**Pull request merged**